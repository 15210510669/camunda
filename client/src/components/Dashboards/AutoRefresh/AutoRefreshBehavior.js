/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';

export default class AutoRefreshBehavior extends React.Component {
  render() {
    return null;
  }

  componentDidMount() {
    this.timer = setInterval(this.props.loadReportData, this.props.interval);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.interval !== this.props.interval ||
      prevProps.loadReportData !== this.props.loadReportData
    ) {
      clearInterval(this.timer);
      this.timer = setInterval(this.props.loadReportData, this.props.interval);
    }
  }
}
