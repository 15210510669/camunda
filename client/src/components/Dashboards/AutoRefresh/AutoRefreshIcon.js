import React from 'react';

import {Icon} from 'components';

import './AutoRefreshIcon.css';

const imageSize = 16;
const center = imageSize / 2;
const radius = center - 1;

export default class AutoRefreshIcon extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      animationStarted: Date.now()
    };
  }

  animate = () => {
    if (this.props.interval && this.path) {
      const progress = ((Date.now() - this.state.animationStarted) / this.props.interval) % 1;
      const angle = Math.PI * (progress * 2 - 0.5);

      const endX = center + radius * Math.cos(angle);
      const endY = center + radius * Math.sin(angle);

      this.path.setAttribute(
        'd',
        `M ${center} ${center - radius} A ${radius} ${radius} 0 ${+(
          progress > 0.5
        )} 1 ${endX} ${endY}`
      );
    }

    this.rafId = requestAnimationFrame(this.animate);
  };

  componentDidMount() {
    this.rafId = requestAnimationFrame(this.animate);
  }

  componentWillUnmount() {
    cancelAnimationFrame(this.rafId);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.interval !== this.props.interval) {
      this.setState({animationStarted: Date.now()});
    }
  }

  render = () =>
    !this.props.interval ? (
      <Icon type="autorefresh" />
    ) : (
      <Icon>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox={`0 0 ${imageSize} ${imageSize}`}
          className="AutoRefreshIcon"
        >
          <circle cx={center} cy={center} r={radius} strokeWidth="0" fillOpacity="0.2" />
          <path
            strokeWidth="2"
            ref={path => (this.path = path)}
            className="AutoRefreshIcon__outline"
          />
        </svg>
      </Icon>
    );
}
