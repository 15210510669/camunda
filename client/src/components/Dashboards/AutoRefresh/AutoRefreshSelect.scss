.AutoRefreshSelect {
  .cds--menu-button__trigger--open svg:not(.cds--btn__icon) {
    transform: none;
  }
}
