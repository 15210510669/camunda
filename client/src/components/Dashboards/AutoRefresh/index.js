/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

export {default as AutoRefreshBehavior} from './AutoRefreshBehavior';
export {default as AutoRefreshIcon} from './AutoRefreshIcon';
export {default as AutoRefreshSelect} from './AutoRefreshSelect';
