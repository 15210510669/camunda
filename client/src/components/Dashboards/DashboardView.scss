.dark {
  .fullscreen {
    background-color: var(--grey-darken-13);
  }
}

.fullscreen {
  background-color: var(--silver-lighten-99);
  transition: background-color 0.3s;
  height: inherit;
}

.fullscreen-enabled {
  overflow: auto;
}

.DashboardView {
  height: inherit;
  display: flex;
  flex-direction: column;

  > .content {
    height: calc(100% - 75px);
  }

  .OptimizeReport:hover {
    .Popover__button {
      opacity: 1;
    }
  }

  .share-button .Popover__dialog {
    padding: 30px;
    padding-bottom: 10px;
  }

  &.fullscreen {
    padding: 20px;

    .ReportDetails .modalsButtons {
      display: none;
    }
  }
}
