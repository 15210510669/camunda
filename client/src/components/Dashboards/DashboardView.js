/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import moment from 'moment';
import classnames from 'classnames';
import Fullscreen from 'react-full-screen';
import {Link} from 'react-router-dom';
import {
  checkDeleteConflict,
  toggleEntityCollection,
  createEntity,
  evaluateReport,
  getEntitiesCollections,
  loadEntities
} from 'services';

import {
  Button,
  ShareEntity,
  DashboardRenderer,
  Icon,
  Dropdown,
  Popover,
  ConfirmationModal,
  CollectionsDropdown,
  EditCollectionModal
} from 'components';

import {themed} from 'theme';

import {getSharedDashboard, shareDashboard, revokeDashboardSharing} from './service';

import {DimensionSetter} from './DimensionSetter';
import {AutoRefreshBehavior, AutoRefreshIcon} from './AutoRefresh';

import './DashboardView.scss';

export default themed(
  class DashboardView extends React.Component {
    state = {
      fullScreenActive: false,
      autoRefreshInterval: null,
      creatingCollection: false,
      confirmModalVisible: false,
      conflicts: [],
      collections: [],
      deleteLoading: false
    };

    componentDidMount() {
      this.loadCollections();
    }

    componentWillUnmount = () => {
      if (this.props.theme === 'dark') {
        this.props.toggleTheme();
      }
    };

    changeFullScreen = fullScreenActive => {
      if (this.props.theme === 'dark') {
        this.props.toggleTheme();
      }
      this.setState({fullScreenActive});
    };

    setAutorefresh = timeout => {
      clearInterval(this.timer);
      if (timeout) {
        this.timer = setInterval(this.props.loadDashboard, timeout);
      }
      this.setState({
        autoRefreshInterval: timeout
      });
    };

    showDeleteModal = async () => {
      this.setState({
        confirmModalVisible: true,
        deleteLoading: true
      });
      const {conflictedItems} = await checkDeleteConflict(this.props.id, 'dashboard');
      this.setState({conflicts: conflictedItems, deleteLoading: false});
    };

    getShareTooltip = () => {
      if (!this.props.sharingEnabled) {
        return 'Sharing is disabled per configuration';
      }
      if (!this.props.isAuthorizedToShare) {
        return (
          'You are not authorized to share the dashboard, ' +
          " because you don't have access to all reports on the dashboard!"
        );
      }
      return '';
    };

    autoRefreshOption = (interval, label) => {
      return (
        <Dropdown.Option
          active={this.state.autoRefreshInterval === interval}
          onClick={() => this.setAutorefresh(interval)}
        >
          {label}
        </Dropdown.Option>
      );
    };

    openEditCollectionModal = () => {
      this.setState({creatingCollection: true});
    };

    loadCollections = async () => {
      this.setState({collections: await loadEntities('collection', 'created')});
    };

    closeConfirmModal = () => {
      this.setState({
        confirmModalVisible: false
      });
    };

    createCollection = async name => {
      await createEntity('collection', {name, data: {entities: [this.props.id]}});
      await this.loadCollections();
      this.setState({creatingCollection: false});
    };

    render() {
      const {
        id,
        name,
        lastModifier,
        lastModified,
        isAuthorizedToShare,
        sharingEnabled,
        reports
      } = this.props;

      const {
        collections,
        fullScreenActive,
        autoRefreshInterval,
        confirmModalVisible,
        conflicts,
        deleteLoading,
        creatingCollection
      } = this.state;

      const dashboardCollections = getEntitiesCollections(collections)[id];

      return (
        <Fullscreen enabled={fullScreenActive} onChange={this.changeFullScreen}>
          <div
            className={classnames('DashboardView', {
              fullscreen: fullScreenActive
            })}
          >
            <div className="header">
              <div className="head">
                <div className="name-container">
                  <h1 className="name">{name}</h1>
                </div>
                <div className="tools">
                  {!fullScreenActive && (
                    <React.Fragment>
                      <Link
                        className="tool-button edit-button"
                        to={`/dashboard/${id}/edit`}
                        onClick={() => this.setAutorefresh(null)}
                      >
                        <Button>
                          <Icon type="edit" />
                          Edit
                        </Button>
                      </Link>
                      <Button onClick={this.showDeleteModal} className="tool-button delete-button">
                        <Icon type="delete" />
                        Delete
                      </Button>
                      <Popover
                        className="tool-button share-button"
                        icon="share"
                        title="Share"
                        disabled={!sharingEnabled || !isAuthorizedToShare}
                        tooltip={this.getShareTooltip()}
                      >
                        <ShareEntity
                          type="dashboard"
                          resourceId={id}
                          shareEntity={shareDashboard}
                          revokeEntitySharing={revokeDashboardSharing}
                          getSharedEntity={getSharedDashboard}
                        />
                      </Popover>
                    </React.Fragment>
                  )}
                  {fullScreenActive && (
                    <Button onClick={this.props.toggleTheme} className="tool-button theme-toggle">
                      Toggle Theme
                    </Button>
                  )}
                  <Button
                    onClick={() => this.changeFullScreen(!this.state.fullScreenActive)}
                    className="tool-button fullscreen-button"
                  >
                    <Icon type={fullScreenActive ? 'exit-fullscreen' : 'fullscreen'} />
                    {fullScreenActive ? ' Leave' : ' Enter'} Fullscreen
                  </Button>
                  <Dropdown
                    label={
                      <React.Fragment>
                        <AutoRefreshIcon interval={autoRefreshInterval} /> Auto Refresh
                      </React.Fragment>
                    }
                    active={!!autoRefreshInterval}
                  >
                    {this.autoRefreshOption(null, 'Off')}
                    {this.autoRefreshOption(1 * 60 * 1000, '1 Minute')}
                    {this.autoRefreshOption(5 * 60 * 1000, '5 Minutes')}
                    {this.autoRefreshOption(10 * 60 * 1000, '10 Minutes')}
                    {this.autoRefreshOption(15 * 60 * 1000, '15 Minutes')}
                    {this.autoRefreshOption(30 * 60 * 1000, '30 Minutes')}
                    {this.autoRefreshOption(60 * 60 * 1000, '60 Minutes')}
                  </Dropdown>
                </div>
              </div>
              <div className="subHead">
                <div className="metadata">
                  Last modified {moment(lastModified).format('lll')} by {lastModifier}
                </div>
                <CollectionsDropdown
                  entity={{id}}
                  collections={collections}
                  toggleEntityCollection={toggleEntityCollection(this.loadCollections)}
                  entityCollections={dashboardCollections}
                  setCollectionToUpdate={this.openEditCollectionModal}
                />
              </div>
            </div>
            <ConfirmationModal
              open={confirmModalVisible}
              onClose={this.closeConfirmModal}
              onConfirm={this.props.deleteDashboard}
              conflict={{type: 'Delete', items: conflicts}}
              entityName={name}
              loading={deleteLoading}
            />
            <DashboardRenderer
              loadReport={evaluateReport}
              reports={reports}
              reportAddons={
                autoRefreshInterval && [
                  <AutoRefreshBehavior key="autorefresh" interval={autoRefreshInterval} />
                ]
              }
            >
              <DimensionSetter reports={reports} />
            </DashboardRenderer>
          </div>
          {creatingCollection && (
            <EditCollectionModal
              collection={{}}
              onClose={() => this.setState({creatingCollection: false})}
              onConfirm={this.createCollection}
            />
          )}
        </Fullscreen>
      );
    }
  }
);
