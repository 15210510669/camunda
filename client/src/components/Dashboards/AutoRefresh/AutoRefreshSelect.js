/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useState, useEffect, useRef} from 'react';
import {Button, MenuItemSelectable, Tooltip, Menu} from '@carbon/react';
import {ChevronDown, UpdateNow} from '@carbon/icons-react';
import {useAttachedMenu} from '@camunda/camunda-optimize-composite-components';
import classnames from 'classnames';

import {t} from 'translation';

import AutoRefreshIcon from './AutoRefreshIcon';

import './AutoRefreshSelect.scss';

export default function AutoRefreshSelect({refreshRateMs, onRefresh, onChange}) {
  const [autoRefreshHandle, setAutoRefreshHandle] = useState(() =>
    refreshRateMs && onRefresh ? setInterval(onRefresh, refreshRateMs) : null
  );
  const menuRef = useRef(null);
  const triggerRef = useRef(null);
  const [width, setWidth] = useState(0);
  const {
    open,
    x,
    y,
    handleClick: hookOnClick,
    handleMousedown,
    handleClose,
  } = useAttachedMenu(triggerRef);

  function setAutorefresh(newRefreshRateMs) {
    clearInterval(autoRefreshHandle);
    if (newRefreshRateMs && onRefresh) {
      setAutoRefreshHandle(setInterval(onRefresh, newRefreshRateMs));
    }
    onChange(newRefreshRateMs);
  }

  function handleClick() {
    if (triggerRef.current) {
      const {width} = triggerRef.current.getBoundingClientRect();
      setWidth(width);
      hookOnClick();
    }
  }

  function handleOpen() {
    if (menuRef.current) {
      menuRef.current.style.width = `${width}px`;
      menuRef.current.style.maxWidth = 'unset';
    }
  }

  useEffect(() => {
    return () => {
      clearInterval(autoRefreshHandle);
    };
  }, [autoRefreshHandle]);

  const menuItems = [
    {value: 'off', label: t('common.off')},
    {value: '1', label: `1 ${t('common.unit.minute.label')}`},
    {value: '5', label: `5 ${t('common.unit.minute.label-plural')}`},
    {value: '10', label: `10 ${t('common.unit.minute.label-plural')}`},
    {value: '15', label: `15 ${t('common.unit.minute.label-plural')}`},
    {value: '30', label: `30 ${t('common.unit.minute.label-plural')}`},
    {value: '60', label: `60 ${t('common.unit.minute.label-plural')}`},
  ];

  return (
    <Tooltip label={t('dashboard.autoRefresh')} className="cds--icon-tooltip" leaveDelayMs={50}>
      <div className="AutoRefreshSelect">
        <Button
          className={classnames('cds--menu-button__trigger', {
            'cds--menu-button__trigger--open': open,
          })}
          ref={triggerRef}
          kind="ghost"
          iconDescription={t('dashboard.autoRefresh')}
          renderIcon={ChevronDown}
          aria-haspopup
          aria-expanded={open}
          onClick={handleClick}
          onMouseDown={handleMousedown}
          aria-controls={open ? 'autoRefreshSelect' : undefined}
        >
          {onRefresh ? <AutoRefreshIcon interval={refreshRateMs} /> : <UpdateNow />}
        </Button>
        <Menu
          ref={menuRef}
          target={triggerRef.current}
          id={'autoRefreshSelect'}
          label={t('dashboard.autoRefresh')}
          open={open}
          onClose={handleClose}
          onOpen={handleOpen}
          x={x}
          y={y}
        >
          {menuItems.map(({value, label}) => {
            const currentValue = !refreshRateMs ? 'off' : (refreshRateMs / (60 * 1000)).toString();
            return (
              <MenuItemSelectable
                key={value}
                value={value}
                label={label}
                selected={value === currentValue}
                onChange={() => setAutorefresh(value === 'off' ? null : value * 60 * 1000)}
              />
            );
          })}
        </Menu>
      </div>
    </Tooltip>
  );
}
