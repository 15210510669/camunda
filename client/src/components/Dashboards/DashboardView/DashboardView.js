import React from 'react';

import {DashboardReport} from '../DashboardReport';
import {DashboardObject} from '../DashboardObject';

import './DashboardView.css';

const columns = 18;
const tileAspectRatio = 1;
const tileMargin = 8;

export default class DashboardView extends React.Component {
  render() {
    return (<div className='DashboardView' ref={node => this.container = node}>
      {this.state && this.props.reports.map((report, idx) =>
        <DashboardObject
          key={idx + '_' + report.id}
          tileDimensions={this.state.tileDimensions}
          {...report.position}
          {...report.dimensions}
        >
        <DashboardReport report={report} tileDimensions={this.state.tileDimensions} addons={this.props.reportAddons || []} />
      </DashboardObject>)}
      {this.state &&
        React.Children.map(this.props.children, child =>
          React.cloneElement(child, {
            container: this.container,
            tileDimensions: this.state.tileDimensions,
            reports: this.props.reports
          })
        )
      }
    </div>);
  }

  componentDidMount() {
    const tileDimensions = this.getTileDimensions();

    this.container.style.width = tileDimensions.outerWidth * columns + 'px';

    this.setState({
      tileDimensions
    });
  }

  getTileDimensions = () => {
    const availableWidth = this.container.clientWidth;
    const outerWidth = ~~(availableWidth / columns); // make sure we are working with round values

    const innerWidth = outerWidth - tileMargin;
    const innerHeight = innerWidth / tileAspectRatio;

    const outerHeight = innerHeight + tileMargin;

    return {outerWidth, innerWidth, outerHeight, innerHeight, columns};
  }
}
