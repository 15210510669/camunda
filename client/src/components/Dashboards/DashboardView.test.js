/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React, {runAllEffects, runAllCleanups} from 'react';
import {shallow} from 'enzyme';
import {useFullScreenHandle} from 'react-full-screen';

import {AlertsDropdown} from 'components';
import {getOptimizeProfile} from 'config';

import {AutoRefreshSelect} from './AutoRefresh';
import {DashboardView} from './DashboardView';

jest.mock('config', () => ({
  getOptimizeProfile: jest.fn().mockReturnValue('platform'),
}));

jest.mock('react-full-screen', () => {
  const handle = {active: false, exit: jest.fn(), enter: jest.fn()};
  return {
    FullScreen: (props) => <div {...props} />,
    useFullScreenHandle: () => handle,
  };
});
jest.mock('./service', () => ({
  getDefaultFilter: () => [],
}));

beforeEach(() => {
  useFullScreenHandle().active = false;
  useFullScreenHandle().enter.mockClear();
  useFullScreenHandle().exit.mockClear();
});

it('should display the key properties of a dashboard', () => {
  const node = shallow(
    <DashboardView
      name="name"
      lastModifier="lastModifier"
      lastModified="2020-11-11T11:11:11.1111+0200"
    />
  );

  expect(node.find('EntityName').prop('children')).toBe('name');
});

it('should provide a link to edit mode in view mode', () => {
  const node = shallow(<DashboardView currentUserRole="editor" />);

  expect(node.find('.edit-button')).toExist();
});

it('should render a sharing popover', async () => {
  const node = shallow(<DashboardView />);

  await runAllEffects();

  expect(node.find('Popover.share-button')).toExist();
});

it('should show sharing popover in cloud environment', async () => {
  getOptimizeProfile.mockReturnValueOnce('cloud');
  const node = shallow(<DashboardView />);

  await runAllEffects();

  expect(node.find('Popover.share-button')).toExist();
});

it('should hide alert dropdown in ccsm environment', async () => {
  getOptimizeProfile.mockReturnValueOnce('ccsm');
  const node = shallow(<DashboardView />);

  await runAllEffects();

  expect(node.find(AlertsDropdown)).not.toExist();
});

it('should enter fullscreen mode', () => {
  const node = shallow(<DashboardView />);

  node.find('.fullscreen-button').simulate('click');

  expect(useFullScreenHandle().enter).toHaveBeenCalled();
});

it('should leave fullscreen mode', () => {
  const node = shallow(<DashboardView />);
  useFullScreenHandle().active = true;

  node.find('.fullscreen-button').simulate('click');

  expect(useFullScreenHandle().exit).toHaveBeenCalled();
});

it('should set auto refresh value and pass it to the refresh behavior addon', () => {
  const node = shallow(<DashboardView />);

  node.find(AutoRefreshSelect).simulate('change', 'number');

  expect(node.find(AutoRefreshSelect).prop('refreshRateMs')).toBe('number');
  expect(node.find('DashboardRenderer').prop('addons')).toMatchSnapshot();
});

it('should pass the refresh interval prop to the refreshSelect and refresh bahavior components', () => {
  const intervalSeconds = 60;
  const node = shallow(<DashboardView refreshRateSeconds={intervalSeconds} />);

  expect(node.find(AutoRefreshSelect).prop('refreshRateMs')).toBe(intervalSeconds * 1000);
  expect(node.find('DashboardRenderer').prop('addons')).toMatchSnapshot();
});

it('should have a toggle theme button that is only visible in fullscreen mode', () => {
  let node = shallow(<DashboardView />);

  expect(node.find('.theme-toggle')).not.toExist();

  useFullScreenHandle().active = true;
  node = shallow(<DashboardView />);

  expect(node.find('.theme-toggle')).toExist();
});

it('should toggle the theme when clicking the toggle theme button', () => {
  const spy = jest.fn();
  useFullScreenHandle().active = true;
  const node = shallow(<DashboardView toggleTheme={spy} />);

  node.find('.theme-toggle').simulate('click');

  expect(spy).toHaveBeenCalled();
});

it('should return to light mode when exiting fullscreen mode', () => {
  const spy = jest.fn();
  useFullScreenHandle().active = true;
  const node = shallow(<DashboardView toggleTheme={spy} theme="dark" />);

  node.find('FullScreen').simulate('change');

  expect(spy).toHaveBeenCalled();
});

it('should return to light mode when the component is unmounted', async () => {
  const spy = jest.fn();
  useFullScreenHandle().active = true;
  shallow(<DashboardView toggleTheme={spy} theme="dark" />);

  await runAllEffects();
  await runAllCleanups();

  expect(spy).toHaveBeenCalled();
});

it('should hide alert dropdown in full screen mode', async () => {
  useFullScreenHandle().active = true;
  const node = shallow(<DashboardView />);

  await runAllEffects();

  expect(node.find(AlertsDropdown)).not.toExist();
});

it('should disable the share button if not authorized', async () => {
  const node = shallow(<DashboardView isAuthorizedToShare={false} sharingEnabled />);

  await runAllEffects();

  const shareButton = node.find('.share-button');
  expect(shareButton).toBeDisabled();
  expect(shareButton.props()).toHaveProperty(
    'tooltip',
    'Sharing forbidden. Missing authorization for contained report.'
  );
});

it('should enable share button if authorized', () => {
  const node = shallow(<DashboardView isAuthorizedToShare sharingEnabled />);

  expect(node.find('.share-button')).not.toBeDisabled();
});

it('should hide edit/delete if the dashboard current user role is not "editor"', () => {
  const node = shallow(<DashboardView currentUserRole="viewer" />);

  expect(node.find('.delete-button')).not.toExist();
  expect(node.find('.edit-button')).not.toExist();
});

it('should show a filters toggle button if filters are available', () => {
  const node = shallow(<DashboardView availableFilters={[{type: 'state'}]} />);

  expect(node.find('.filter-button')).toExist();
});

it('should toggle filters section', () => {
  const node = shallow(<DashboardView availableFilters={[{type: 'state'}]} />);

  expect(node.find('.filter-button')).toHaveProp('active');
  expect(node.find('FiltersView')).toExist();

  node.find('.filter-button').simulate('click');

  expect(node.find('.filter-button').prop('active')).toBe(false);
  expect(node.find('FiltersView')).not.toExist();
});

it('should reset filters when closing the filters section', () => {
  const filter = [{type: 'runningInstancesOnly', data: null}];
  const node = shallow(<DashboardView availableFilters={[{type: 'state'}]} />);

  node.find('FiltersView').prop('setFilter')(filter);

  expect(node.find('FiltersView').prop('filter')).toEqual(filter);
  node.find('.filter-button').simulate('click');
  node.find('.filter-button').simulate('click');
  expect(node.find('FiltersView').prop('filter')).toEqual([]);
});

it('should hide the share button', () => {
  const node = shallow(<DashboardView sharingHidden />);

  expect(node.find('Popover.share-button')).not.toExist();
});
