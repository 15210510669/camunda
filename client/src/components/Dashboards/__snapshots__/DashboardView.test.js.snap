// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add an autorefresh addon when autorefresh mode is active 1`] = `
Array [
  <AutoRefreshBehavior
    interval={3600000}
  />,
  <DiagramScrollLock />,
]
`;
